lib/utils.ts

```diff
@@ -0,0 +1,209 @@
+// eslint-disable-next-line one-var
+import * as appRoot from 'app-root-path'
+import { version } from '../package.json'
+import { ValueID } from 'zwave-js'
+import path = require('path')
+import crypto = require('crypto')
+import { execSync } from 'child_process'
+
+let VERSION: string
+
+export type DeepPartial<T> = {
+	[P in keyof T]?: T[P] extends Array<infer U>
+		? Array<DeepPartial<U>>
+		: T[P] extends ReadonlyArray<infer U>
+		? ReadonlyArray<DeepPartial<U>>
+		: DeepPartial<T[P]>
+}
+
+/**
+ *  Get the base root path to application directory. When we are in a `pkg` environment
+ *  the path of the snapshot is not writable
+ */
+export function getPath(write: boolean): string {
+	if (write && hasProperty(process, 'pkg')) return process.cwd()
+	else return appRoot.toString()
+}
+
+/**
+ * path.join wrapper, the first option can be a boolean and it will automatically fetch the root path
+ * passing the boolean to getPath
+ */
+export function joinPath(write: boolean | string, ...paths: string[]): string {
+	if (typeof write === 'boolean') {
+		write = getPath(write)
+	}
+	return path.join(write, ...paths)
+}
+
+/**
+ * Join props with a `_` and skips undefined props
+ */
+export function joinProps(...props: (string | number)[]): string {
+	props = props || []
+	let ret = props[0].toString() || ''
+	for (let i = 1; i < props.length; i++) {
+		const p = props[i]
+		if (p !== null && p !== undefined && p !== '') {
+			ret += '_' + (typeof p === 'number' ? p.toString() : p)
+		}
+	}
+	return ret
+}
+
+/**
+ * Checks if an object is a valueId, returns error otherwise
+ */
+export function isValueId(v: ValueID): boolean | string {
+	if (typeof v.commandClass !== 'number' || v.commandClass < 0) {
+		return 'invalid `commandClass`'
+	}
+	if (v.endpoint !== undefined && v.endpoint < 0) {
+		return 'invalid `endpoint`'
+	}
+	if (
+		v.property === undefined ||
+		(typeof v.property !== 'string' && typeof v.property !== 'number')
+	) {
+		return 'invalid `property`'
+	}
+	if (
+		v.propertyKey !== undefined &&
+		typeof v.propertyKey !== 'string' &&
+		typeof v.propertyKey !== 'number'
+	) {
+		return 'invalid `propertyKey`'
+	}
+	return true
+}
+
+/**
+ * Deep copy of an object
+ */
+export function copy(obj: any): any {
+	return JSON.parse(JSON.stringify(obj))
+}
```

Comment:
I hope you don't do this too often on large objects. There are more performant ways to deep copy.

Reply:
Like what?